worker/provisioner/provisioner_task.go

```diff
@@ -689,106 +714,357 @@ func (task *provisionerTask) maintainMachines(machines []*apiprovisioner.Machine
+	}
+
+	return pInfo, startInstanceParams, nil
 }
 
 func (task *provisionerTask) startMachine(
 	machine *apiprovisioner.Machine,
 	provisioningInfo *params.ProvisioningInfo,
 	startInstanceParams environs.StartInstanceParams,
 ) error {
-	var result *environs.StartInstanceResult
 	// TODO (jam): 2017-01-19 Should we be setting this earlier in the cycle?
 	if err := machine.SetInstanceStatus(status.Provisioning, "starting", nil); err != nil {
 		logger.Errorf("%v", err)
 	}
-	for attemptsLeft := task.retryStartInstanceStrategy.retryCount; attemptsLeft >= 0; attemptsLeft-- {
+
+	// ToDo ProvisionerParallelization 2017-10-03
+	// Improve the retry loop, newer methodology
+	// Is rate limiting handled correctly?
+
+	// Increase attempt by a multiplier of the number of availability aones
+	// to account for the retries removed from the providers when the provisioner
+	// was parallelized.
+	task.azMachinesMutex.RLock()
+	var attemptsLeft int
+	if len(task.availabilityZoneMachines) > 0 {
+		attemptsLeft = task.retryStartInstanceStrategy.retryCount * len(task.availabilityZoneMachines)
+	} else {
+		// Some providers don't have availability zones
+		attemptsLeft = task.retryStartInstanceStrategy.retryCount
+	}
+	task.azMachinesMutex.RUnlock()
+
+	var result *environs.StartInstanceResult
+	for ; attemptsLeft >= 0; attemptsLeft-- {
 		attemptResult, err := task.broker.StartInstance(startInstanceParams)
 		if err == nil {
 			result = attemptResult
 			break
 		} else if attemptsLeft <= 0 {
 			// Set the state to error, so the machine will be skipped
-			// next time until the error is resolved, but don't return
-			// an error; just keep going with the other machines.
+			// next time until the error is resolved.
 			return task.setErrorStatus("cannot start instance for machine %q: %v", machine, err)
 		}
 
 		retryMsg := fmt.Sprintf("failed to start instance (%s), retrying in %v (%d more attempts)",
 			err.Error(), task.retryStartInstanceStrategy.retryDelay, attemptsLeft)
 		logger.Warningf(retryMsg)
-		if err2 := machine.SetInstanceStatus(status.Provisioning, retryMsg, nil); err2 != nil {
-			logger.Errorf("%v", err2)
+		if err3 := machine.SetInstanceStatus(status.Provisioning, retryMsg, nil); err3 != nil {
+			logger.Errorf("%v", err3)
 		}
 
 		select {
 		case <-task.catacomb.Dying():
 			return task.catacomb.ErrDying()
 		case <-time.After(task.retryStartInstanceStrategy.retryDelay):
 		}
+
+		// If the failure of StartInstance() is due the availability zone,
+		// find a new one to use when retrying.
+		if err == environs.ErrAvailabilityZoneFailed {
+			if err2 := task.markMachineFailedInAZ(machine, startInstanceParams.Placement); err2 != nil {
+				return err2
+			}
+			// ask for a new zone
+			machineZone, err := task.machineAvailabilityZoneDistribution([]*apiprovisioner.Machine{machine})
+			if err != nil {
+				return err
```

Comment:
if we get an error here, we should call setErrorStatus to set the instance status to error

Reply:
done